fdbserver/BlobWorker.actor.cpp

```diff
@@ -104,68 +97,46 @@ struct GranuleMetadata : NonCopyable, ReferenceCounted<GranuleMetadata> {
 	int64_t continueEpoch;
 	int64_t continueSeqno;
 
-	Future<GranuleChangeFeedInfo> assignFuture;
-	Future<Void> fileUpdaterFuture;
 	Promise<Void> resumeSnapshot;
+
+	// used to coordinate granule file updater is done
 	Promise<Void> cancelled;
 	Promise<Void> readable;
 
 	AssignBlobRangeRequest originalReq;
 
-	Future<Void> start(BlobWorkerData* bwData, AssignBlobRangeRequest req) {
-		originalReq = req;
-		assignFuture = persistAssignWorkerRange(bwData, req);
-		fileUpdaterFuture = blobGranuleUpdateFiles(bwData, Reference<GranuleMetadata>::addRef(this));
-
-		return success(assignFuture);
-	}
-
 	void resume() {
 		ASSERT(resumeSnapshot.canBeSet());
 		resumeSnapshot.send(Void());
 	}
-
-	// FIXME: right now there is a dependency because this contains both the actual file/delta data as well as the
-	// metadata (worker futures), so removing this reference from the map doesn't actually cancel the workers. It'd be
-	// better to have this in 2 separate objects, where the granule metadata map has the futures, but the read
-	// queries/file updater/range feed only copy the reference to the file/delta data.
-	Future<Void> cancel(bool dispose) {
-		if (cancelled.canBeSet()) {
-			// Could have been cancelled already by rollback or error in BGUpdateFiles
-			cancelled.send(Void());
-		}
-		assignFuture.cancel();
-		fileUpdaterFuture.cancel();
-
-		if (dispose) {
-			// FIXME: implement dispose!
-			return delay(0.1);
-		}
-		return Future<Void>(Void());
-	}
 };
 
-// for a range that may or may not be set
+// TODO: rename this struct
 struct GranuleRangeMetadata {
 	int64_t lastEpoch;
 	int64_t lastSeqno;
 	Reference<GranuleMetadata> activeMetadata;
 
+	Future<GranuleChangeFeedInfo> assignFuture;
+	Future<Void> fileUpdaterFuture;
+
 	GranuleRangeMetadata() : lastEpoch(0), lastSeqno(0) {}
 	GranuleRangeMetadata(int64_t epoch, int64_t seqno, Reference<GranuleMetadata> activeMetadata)
 	  : lastEpoch(epoch), lastSeqno(seqno), activeMetadata(activeMetadata) {}
 };
 
-struct BlobWorkerData {
+struct BlobWorkerData : NonCopyable, ReferenceCounted<BlobWorkerData> {
```

Comment:
Oh man I just realized we have another reference cycle here :(
When we want to destroy a blob worker's data, the actors in the granuleMetadata map have a copy of the BlobWorkerData reference :(
We don't have to solve that in this PR though, but it does mean that when we stop a blob worker, this won't get cleaned up properly, and we should at least add a FIXME.
The fix would be to not pass Reference<BlobWorkerData> to any of the forever-looping actors that BlobWorkerData "owns", and just pass a copy of the relevant fields. Most of them only use db anyway i think.
Your changes do fix it for cleaning stuff up on revoke though which is the more important case.